### DISCUSSION TITLE
How to add Header to every route in a Router such that there is a Header field visible for every route within the APIRouter when Swagger UI opens

### DISCUSSION THREAD

**Discussion by jayashankarjayan** [fastapi/fastapi#6086] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import json
from fastapi import FastAPI, Header, Request
from fastapi.openapi.utils import get_openapi

app = FastAPI()


@app.get("/items/")
async def read_items(x_tenant_id: str = Header()):
    return [{"hello": "world"}]


@app.get("/users")
async def read_users(request: Request):
    return [{"tag": "@jayashankarjayan"}]


def custom_openapi():
    header = {
            "required": True,
            "schema": {
              "title": "X-Custom-Id",
              "type": "string"
            },
            "name": "x-custom-id",
            "in": "header"
          }
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title="Custom title",
        version="2.5.0",
        description="This is a very custom OpenAPI schema",
        routes=app.routes,
    )
    openapi_schema["info"]["x-logo"] = {
        "url": "https://fastapi.tiangolo.com/img/logo-margin/logo-teal.png"
    }
    app.openapi_schema = openapi_schema

    paths = openapi_schema["paths"]
    for url, methods in paths.items():
        for method in methods:
            if methods[method].get("parameters"):
                methods[method]["parameters"].append(header)
            else:
                methods[method]["parameters"] = [header]

    return app.openapi_schema


app.openapi = custom_openapi
```

### Description

I have a requirement in my application where all my APIs spread across multiple routers are required to have custom headers, eg: `x-custom-header`. I want these headers to be visible as fields on the Swagger UI as well so that a user who accesses this Swagger UI can use the API endpoints by providing valid values to the headers.

I have two options -
1. I add the header as an argument to every single route individually. This introduces a lot of duplication as I have to repeat the same arguments over multiple routes.
2. I modify the OpenAPI Schema (as seen in the above code snippet) and add the header across all routes at one go. While this works, it seems more like a somewhat ugly hack.

I would like to know that is there a better, elegant option than the aforementioned one which I can use to get header fields in my Swagger UI while avoiding duplication in my code and keeping it elegant.

All help is appreciated. Thanks!

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.86.0

### Python Version

3.8.9

### Additional Context

<img width="1509" alt="image" src="https://user-images.githubusercontent.com/20151948/207940178-8cc90f2a-f697-4177-bd5e-d749229acaf3.png">

This is how I would like the custom header to appear without actually mentioning it explicitly on the route itself (the result seen in the screenshot can be replicated using the code snippet shared above)

Kindly note that there could be multiple such custom headers that I want to see on my Swagger UI

**Comment by iudeen** [comment]
You can use a dependency

**Comment by jayashankarjayan** [comment, marked as answer]
Thank you @iudeen . I was able use the `dependencies` argument to the APIRouter and add a dependency and get it working.

Following is the implementation for reference to anyone who comes across such a requirement

```from fastapi import Depends, FastAPI, Header, APIRouter, Request

app = FastAPI()

def custom_headers(x_header_1: str = Header(), x_header_2 = Header()):
    return "hey there"

router = APIRouter(dependencies=[Depends(custom_headers)])

@router.get("/query-checker/")
async def read_query_check(request: Request):
    print(request.headers["X-Header-1"])
    return {"hello": "world"}

app.include_router(router)```

On another note, @iudeen is it possible to access the value returned by the dependency?

**Comment by iudeen** [comment]
Sorry, I was not able to provide you with an example earlier, I was traveling. But great you have done it the right way!

And to answer your question of passing results back into different route functions, you can either add dependency to each of the functions, and the results will get passed on. Or, you can try setting a custom attribute to `Request` object.

**Comment by jayashankarjayan** [comment]
Thank you!

As for setting a custom attribute to the `Request` object, do you mean using the `.state` property of request? In the code shared above I have used a `request` argument to access the headers. This argument would have to be added to every route for me to be able to access the header (or `state` property).  

My goal is to see if we can eliminate the redundant declaration of the same argument across all the routes. I agree that in case of common arguments, we could use class based dependencies and call them on a per-route level.

To better phrase my question, is it possible to access, in an individual route, an object returned by a dependency that is being used at a router level using the `dependencies` argument, without introducing an additional argument to the routes within the router.

**Comment by iudeen** [comment]
@jayashankarjayan no, there is no straight forward solution. 

There is one way using ContexVar, but that approach may be too complex. You can read more on that [here](https://stackoverflow.com/questions/57204499/is-there-a-fastapi-way-to-access-current-request-data-globally)

**Comment by jayashankarjayan** [comment]
I see. I'd rather not go the ContextCar route. It's somewhat of a messy solution. Thanks you for the help. I'll now close this issue

**Comment by iudeen** [comment]
Thanks for closing the issue! 😁
